rewrite-core/src/test/java/org/openrewrite/AddToGitignoreTest.java

```diff
@@ -0,0 +1,258 @@
+              # OS files
+              .DS_Store
+              Thumbs.db
+              """,
+            spec -> spec.path(".gitignore")
+          )
+        );
+    }
+
+    @Test
+    void handleWindowsLineEndings() {
+        rewriteRun(
+          spec -> spec.recipe(new AddToGitignore("*.log")),
+          text(
+            "*.tmp\r\n.DS_Store\r\n",
+            "*.tmp\r\n.DS_Store\r\n\r\n*.log",
+            spec -> spec.path(".gitignore")
+          )
+        );
+    }
+
+    @Test
+    void handleNestedGitignore() {
+        rewriteRun(
+          spec -> spec.recipe(new AddToGitignore("local.properties")),
+          text(
+            """
+              *.class
+              """,
+            """
+              *.class
+              
+              local.properties
+              """,
+            spec -> spec.path("src/.gitignore")
+          )
+        );
+    }
+
+    @Test
+    void normalizeAndDeduplicateEntries() {
+        rewriteRun(
+          spec -> spec.recipe(new AddToGitignore("/build/\nbuild\n/build")),
+          text(
+            """
+              target/
+              """,
+            """
+              target/
+              
+              /build/
+              """,
+            spec -> spec.path(".gitignore")
+          )
+        );
+    }
+
+    @Test
+    void handleNegationPatterns() {
+        rewriteRun(
+          spec -> spec.recipe(new AddToGitignore("*.log\n!important.log")),
+          text(
+            """
+              *.tmp
+              """,
+            """
+              *.tmp
+              
+              *.log
+              !important.log
+              """,
+            spec -> spec.path(".gitignore")
+          )
+        );
+    }
+
+    @Test
+    void doNotAddDuplicateNegations() {
+        rewriteRun(
+          spec -> spec.recipe(new AddToGitignore("!important.log")),
+          text(
+            """
+              *.log
+              !important.log
+              """,
+            spec -> spec.path(".gitignore")
+          )
+        );
+    }
+
+    @Test
+    void emptyLinesInInput() {
+        rewriteRun(
+          spec -> spec.recipe(new AddToGitignore("\n\n*.log\n\n*.tmp\n\n")),
+          text(
+            """
+              target/
+              """,
+            """
+              target/
+              
+              *.log
+              *.tmp
+              """,
+            spec -> spec.path(".gitignore")
+          )
+        );
+    }
+
+    @Test
+    void doNotChangeWhenEntriesExistInDifferentOrder() {
+        rewriteRun(
+          spec -> spec.recipe(new AddToGitignore("*.log\ntarget/\n.DS_Store\nbuild/")),
+          text(
+            """
+              # Build directories
+              build/
+              target/
+              
+              # Logs
+              *.log
+              
+              # OS files
+              .DS_Store
+              """,
+            spec -> spec.path(".gitignore")
+          )
+        );
+    }
+}
```

Comment:
It might be nice to avoid adding entries already matched by existing patterns.
```suggestion
    }

    @Test
    void doNotAddSuperfluousEntries() {
        rewriteRun(
          spec -> spec.recipe(new AddToGitignore("foo.tmp")),
          text(
            """
              *.tmp
              """,
            spec -> spec.path(".gitignore")
          )
        );
    }
}
```